### ISSUE TITLE
General framework for dataset, format-agnostic and compatible with NGFF

### ISSUE THREAD

**Issue by jdeschamps** [CAREamics/careamics#292]
## Description

> [!IMPORTANT]
> Summary: replacing current datasets by a general framework compatible with NGFF file formats.

Currently, we have accumulated a lot of technical debt on our datasets for the sake of simplifying the code. We have ended up with two different types of datasets (InMemory and PathIterable), which have different patching strategies (sequential vs random) and are different types of datasets. Furthermore, the train and prediction datasets have been split because of the tiling in the prediction.

This creates a lot of complexity and different behaviours depending on hyperparameters. The fact that every time we discuss datasets we are confused about which one does what is a bad sign!

To allow additional file formats, we also provide parameters to pass extensions and file reader callables. While this is a nice idea, it is very limiting and will not scale well.

As we are moving towards compatibility with NGFF, we need a better and more general dataset system that can be compatible with multiple back-ends. In addition, this would help third-parties implement their own file format. 

### Features
- Dataset
    - Tiff, Zarr, numpy array
    - Optional targets
    - Add a maximum number of iterations
- Patcher
    - Exclude background coordinates (here or in sampler?)
    - Random patching for training
    - Save train / val coordinates for reproducibility?
    - Seed-able
    - Possibility to fix coordinates (sequential?) for validation
    - Multi-file patching
- Sampler
- Data statistics
    - Welford mean/std
    - Quantile normalization
    - No normalization
    - Pre-computed
- Metadata extractor
    - returns shape of each file for a number of files
    - if metadata is unavailable, defaults to reading the file
- Augmentations
    - Normalization handling
    - Random rotate 90 degrees
    - Random X or Y flip
    - Add noise

### Which part of the code?

- https://github.com/CAREamics/careamics/blob/main/src/careamics/dataset/in_memory_dataset.py
- https://github.com/CAREamics/careamics/blob/main/src/careamics/dataset/iterable_dataset.py
- other datasets in https://github.com/CAREamics/careamics/tree/main/src/careamics/dataset

## Potential solutions

@veegalinova and @CatEek have worked out a potential architecture, please post the schematic here for reference once we will be more confident this is the way to go.

## Relevant issues

https://github.com/CAREamics/careamics/issues/293: Exclude background patches during training.
https://github.com/CAREamics/careamics/issues/282: Why the difference between patching (random vs sequential)?
https://github.com/CAREamics/careamics/issues/297: Should we save train/val patches info for reproducibility?
https://github.com/CAREamics/careamics/issues/315: cryoCARE compatibility, it uses memory-mapped data
https://github.com/CAREamics/careamics/issues/303: Quantile normalization
https://github.com/CAREamics/careamics/issues/342: Add validation dataloader parameters
https://github.com/CAREamics/careamics/issues/343: Validation patches are random with `PathIterable`
https://github.com/CAREamics/careamics/issues/358: Stitching impossible if tiles not ordered by image
https://github.com/CAREamics/careamics/issues/364: Turn off normalization if data already normalized
#293: Exclude background patches during training
#167: Adapt TTA from training configuration

**Comment by melisande-c** [comment]
Idea to return dicts from the Dataset, to allow the inclusion of additional items required for different algorithms more easily:

https://github.com/CAREamics/careamics/pull/420#discussion_r1971315755

**Comment by jdeschamps** [comment]
Let's consider this closed and integrated (will be default in v0.2)
